implementation/alpha/node/node.ino

```diff
@@ -404,69 +437,85 @@ void increment_packet_age(uint32_t time) {
-
-        float reading_f[NUM_SENSORS];
-        read_temperatures(reading_f);
-
-        for (int i = 0; i < NUM_SENSORS; i++) {
-            new_packet.reading[i] = compress_float(reading_f[i]);
-        }
-
-        // Clear space by deleting older packets
-        if (packet_queue.size == QUEUE_SIZE_MAX) {
-            packet_queue.pop();
-        }
-
-        packet_queue.push(new_packet);
-
-        if (PRINT_DEBUG) {
-            Serial.print("Packet created: ");
-            Serial.println(packet_number);
-            print_packet(new_packet);
-        }
-          
-        
-        packet_number++;
-
-        last_reading_time = correct_millis();
-    } else if (my_id > 128) {
-        health_packet_generate();
+    switch (current_state) {
+        case Reading:
+            data_packet_generate();
+            break;
+        case HealthPacket:
+            health_packet_generate();
+            break;
+        case Transmit:
+            loop_tx();
+            transmit_valid = false;  // need to generate new tx time
+            break;
+        case Receive:
+            loop_rx();
+            break;
     }
-
-    // Set frequency
-    float expected_frequency = 0.0f;
-    if (packet_queue.size > 0 && my_data.parent != NO_ID) {
-        expected_frequency = parent_data.rx_frequency;
-    } else {
-        expected_frequency = my_data.rx_frequency;
+    // For debug purposes
+    blink_led_periodically();
+    // Unified sleep
+    // Generate new transmit time if packet_queue is not empty and we just
+    // transmitted
+    if (packet_queue.size > 0 &&
+        (current_state == Transmit || !transmit_valid)) {
+        current_tx_sleep = expovariate(TX_RATE);
+        next_transmit = correct_millis() + current_tx_sleep;
+        transmit_valid = true;
     }
-    if (current_frequency != expected_frequency) {
-        rf69.setFrequency(expected_frequency);
-        current_frequency = expected_frequency;
+    // Generate new receive time if there's no sensors (isRelay)
+    if (!my_data.has_sensor && current_state == Receive) {
+        current_rx_sleep = expovariate(RX_RATE);
+        next_receive = correct_millis() + current_rx_sleep;
     }
-
-    if (packet_queue.size > 0 && my_data.parent != NO_ID) {
-        loop_tx();
-    } else {
-        loop_rx();
+    // Generate new reading generation time if there are sensors and we just
+    // generated a data packet
+    if (my_data.has_sensor && current_state == Reading) {
+        next_reading = correct_millis() + PACKET_PERIOD;
     }
+    // Generate new health packet generation time if is relay and we just
+    // generated a health packet
+    if (!my_data.has_sensor && current_state == HealthPacket) {
+        next_health = correct_millis() + HEALTH_PACKET_PERIOD;
+    }
+    State next_state;
+    uint32_t next_time;
+    if (my_data.has_sensor) {
+        // transmit, or generate data
+        if (next_transmit - correct_millis() < next_reading - correct_millis()) {
```

Comment:
Looks like there is a bug if next_transmit and next_reading are equal (or maybe even just sufficiently close?).
Say they are both 10, millis() = 10 also.
10 - 10 < 10 - 10 is false, so we do else block (reading)
Time passes, let's say 10 ms, so now millis() = 20, while next_reading = 10 + 1000 or something = 1010
We get back to the if statement
next_transmit = 10
next_reading = 1010
millis = 20
10 - 20 < 1010 - 20 is false because 10 - 20 = huge number
transmission doesn't happen, potentially forever?